src/core/utils/flash.cpp

```diff
@@ -0,0 +1,266 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "flash.hpp"
+
+#include <stdio.h>
+
+#include <openthread/platform/flash.h>
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+
+namespace ot {
+
+const uint32_t ot::Flash::sSwapActive;
+const uint32_t ot::Flash::sSwapInactive;
+
+void Flash::Init(void)
+{
+    RecordHeader record;
+
+    otPlatFlashInit(&GetInstance());
+
+    mSwapSize = otPlatFlashGetSwapSize(&GetInstance());
+
+    for (mSwapIndex = 0;; mSwapIndex++)
+    {
+        uint32_t swapMarker;
+
+        if (mSwapIndex >= 2)
+        {
+            Wipe();
+            ExitNow();
+        }
+
+        otPlatFlashRead(&GetInstance(), mSwapIndex, 0, &swapMarker, sizeof(swapMarker));
+
+        if (swapMarker == sSwapActive)
+        {
+            break;
+        }
+    }
+
+    for (mSwapUsed = kSwapMarkerSize; mSwapUsed < mSwapSize; mSwapUsed += record.GetSize())
```

Comment:
This loop does not guarantee that `otPlatFlashRead` will not read past the end of the swap block. Maybe `mSwapUsed < mSwapSize - sizeof(record)`?